gpl: Factor out graphics use in gpl

hongted opened the pull request.

maliberty commented on src/gpl/include/gpl/Replace.h:
```diff
@@ -46,13 +48,21 @@ using Clusters = std::vector<Cluster>;
+  // Use the following class as a template for graphics interface.
+  //
+  // Note: no ownership is transfered as the object will create a new
+  // graphics boject of the same class.
```
```suggestion
  // graphics object of the same class.
```

maliberty commented on src/gpl/src/AbstractGraphics.cpp:
```diff
@@ -0,0 +1,10 @@
+
+namespace gpl {
+
+AbstractGraphics::~AbstractGraphics() = default;
```
Why not put this in the header?

maliberty replied:
There are likely many interfaces that don't follow this rule.  Since we don't use shared libraries it has never mattered but it could in the future so it is a good idea.

maliberty commented:
Please don't force push after code review has started as it breaks the "changes since last review" feature in GH (the last reviewed commit is elimiated).

maliberty commented on src/gpl/include/gpl/AbstractGraphics.h:
```diff
@@ -0,0 +1,138 @@
+  virtual ~AbstractGraphics();
+
+  // Create a new object of the same class.
+  virtual std::unique_ptr<AbstractGraphics> MakeNew(utl::Logger* logger) const
```
Why are we making copies of the graphics object rather than share one?

hongted commented:
> Please don't force push after code review has started as it breaks the "changes since last review" feature in GH (the last reviewed commit is elimiated).

Understood, I'll do so in the future.

Also, I don't know if it's the same as "changes since last review",  you can see the updates after a force-push via the "compare" link in this discussion.  It does include changes due to the rebase and not just the commit though.
